lark/tree_templates.py

```diff
@@ -0,0 +1,154 @@
+"""This module defines utilities for matching and translation tree templates.
+
+A tree templates is a tree that contains nodes that are template variables.
+
+"""
+
+from typing import Union, Optional, Mapping
+
+from lark import Tree, Transformer
+
+TreeOrCode = Union[Tree, str]
+
+class TemplateConf:
+    """Template Configuration
+
+    Allows customization for different uses of Template
+    """
+
+    def __init__(self, parse=None):
+        self._parse = parse
+
+
+    def test_var(self, var: Union[Tree, str]) -> Optional[str]:
+        """Given a tree node, if it is a template variable return its name. Otherwise, return None.
+
+        This method may be overridden for customization
+
+        Parameters:
+            var: Tree | str - The tree node to test
+
+        """
+        if isinstance(var, str) and var.startswith('$'):
+            return var.lstrip('$')
+
+        if isinstance(var, Tree) and var.data == 'var' and var.children[0].startswith('$'):
+            return var.children[0].lstrip('$')
+
+
+    def _get_tree(self, template: TreeOrCode):
+        if isinstance(template, str):
+            assert self._parse
+            template = self._parse(template)
+
+        assert isinstance(template, Tree)
+        return template
+
+    def __call__(self, template):
+        return Template(template, conf=self)
+
+    def _match_tree_template(self, template, tree):
+        template_var = self.test_var(template)
+        if template_var:
+            return {template_var: tree}
+
+        if isinstance(template, str):
+            if template == tree:
+                return {}
+            return
+
+        assert isinstance(template, Tree), template
+
+        if template.data == tree.data and len(template.children) == len(tree.children):
+            res = {}
+            for t1, t2 in zip(template.children, tree.children):
+                matches = self._match_tree_template(t1, t2)
+                if matches is None:
+                    return
+
+                res.update(matches)
+
+            return res
+
+
+
+class _ReplaceVars(Transformer):
+    def __init__(self, conf, vars):
+        self.conf = conf

```

Comment:
These should also have an `_` underscore

Reply:
You mean `self._conf`?

Reply:
Yeah. In case the grammar has a rule `conf`